File: python/src/threads/tools/lsp_driver.py
"""The lsp tool's in-sandbox driver. The host uploads this file and runs it with
the sandbox image's python3, which may be as old as 3.8, so it keeps to that language level and
imports nothing from threads. It starts the language server, opens one file, asks one question
and prints one JSON line: {"result": ...} or {"unavailable": reason}.

    python3 lsp_driver.py '<server argv json>' <root> <path> <language> <operation> [line char]
"""

from __future__ import annotations

import asyncio
import json
import os
import shutil
import sys
from pathlib import Path
from typing import TYPE_CHECKING, Final

if TYPE_CHECKING:
    from typing import TypeAlias

Json: TypeAlias = "dict[str, Json] | list[Json] | str | int | float | bool | None"  # noqa: UP040 - runs on Python 3.8
Message: TypeAlias = "dict[str, Json]"  # noqa: UP040 - runs on Python 3.8
# Before 3.11, asyncio's timeout is its own class.
Timeout: Final = asyncio.TimeoutError

READY_S: Final = 30.0
SEARCH: Final = "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
"""Where a bare server command is looked up. The driver runs with the tool env, which has no
PATH, so the default search (/bin:/usr/bin) would miss servers in /usr/local/bin, where pip and
npm put them. Fixed system dirs, never a host PATH."""
QUIET_S: Final = 1.0
REQUESTS: Final = {
    "definition": "textDocument/definition",
    "references": "textDocument/references",
    "hover": "textDocument/hover",
    "symbols": "textDocument/documentSymbol",
}


class Server:
    def __init__(self, proc: asyncio.subprocess.Process) -> None:
        if proc.stdin is None or proc.stdout is None:
            raise AssertionError("pipes were requested")
        self.proc: asyncio.subprocess.Process = proc
        self.stdin: asyncio.StreamWriter = proc.stdin
        self.inbox: asyncio.Queue[Message] = asyncio.Queue()
        self.reading: asyncio.Future[None] = asyncio.ensure_future(self._read(proc.stdout))

    async def _read(self, stdout: asyncio.StreamReader) -> None:
        while True:
            length = 0
            line = await stdout.readline()
            while line not in (b"\r\n", b"\n", b""):
                name, _, value = line.decode("ascii", "replace").partition(":")
                if name.strip().lower() == "content-length":
                    length = int(value)
                line = await stdout.readline()
            if not line:
                return
            message: Json = json.loads(await stdout.readexactly(length))
            if isinstance(message, dict):
                await self.inbox.put(message)

    async def send(self, message: Message) -> None:
        body = json.dumps({"jsonrpc": "2.0", **message}).encode()
        self.stdin.write(b"Content-Length: %d\r\n\r\n" % len(body) + body)
        await self.stdin.drain()

    async def _next(self, wait: float) -> Message | None:
        try:
            return await asyncio.wait_for(self.inbox.get(), wait)
        except Timeout:
            return None

    async def answer(self, n: int) -> Message | None:
        """The response to request `n`, or None if the server stays silent."""
        loop = asyncio.get_running_loop()
        until = loop.time() + READY_S
        while (message := await self._next(until - loop.time())) is not None:
            if message.get("id") == n and "method" not in message:
                return message
        return None

    async def diagnostics(self, uri: str) -> Json:
        """The last publishDiagnostics for the file, once the server is quiet a moment."""
        latest: Json = None
        wait = READY_S
        while (message := await self._next(wait)) is not None:
            params = message.get("params")
            method = message.get("method")
            published = method == "textDocument/publishDiagnostics" and isinstance(params, dict)
            if published and isinstance(params, dict) and params.get("uri") == uri:
                latest, wait = params.get("diagnostics"), QUIET_S
        return latest

    async def close(self) -> None:
        try:
            await self.send({"id": 0, "method": "shutdown"})
            await self.send({"method": "exit"})
            await asyncio.wait_for(self.proc.wait(), QUIET_S)
        except (OSError, Timeout):
            self.proc.kill()
        self.reading.cancel()


async def ask(server: Server, argv: list[str]) -> Json:
    _, root, path, language, operation = argv[:5]
    caps: Json = {"textDocument": {"publishDiagnostics": {}, "hover": {}, "definition": {}}}
    init: Message = {"processId": None, "rootUri": Path(root).as_uri(), "capabilities": caps}
    await server.send({"id": 1, "method": "initialize", "params": init})
    if await server.answer(1) is None:
        return {"unavailable": "the language server did not answer initialize"}
    await server.send({"method": "initialized", "params": {}})
    uri = Path(path).as_uri()
    text = Path(path).read_text(encoding="utf-8", errors="replace")
    doc: Json = {"uri": uri, "languageId": language, "version": 1, "text": text}
    await server.send({"method": "textDocument/didOpen", "params": {"textDocument": doc}})
    if operation == "diagnostics":
        return {"result": await server.diagnostics(uri)}
    params: Message = {"textDocument": {"uri": uri}}
    if len(argv) > 5:  # noqa: PLR2004 - line and character follow the operation
        params["position"] = {"line": int(argv[5]) - 1, "character": int(argv[6]) - 1}
    if operation == "references":
        params["context"] = {"includeDeclaration": True}
    await server.send({"id": 2, "method": REQUESTS[operation], "params": params})
    got = await server.answer(2)
    if got is None:
        return {"unavailable": "the language server did not answer " + operation}
    return {"result": got.get("result", got.get("error"))}


async def main(argv: list[str]) -> Json:
    parsed: Json = json.loads(argv[0])
    command = [a for a in parsed if isinstance(a, str)] if isinstance(parsed, list) else []
    if not command:
        return {"unavailable": "no language server command"}
    command[0] = shutil.which(command[0], path=os.environ.get("PATH", SEARCH)) or command[0]
    try:
        proc = await asyncio.create_subprocess_exec(
            *command,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            cwd=argv[1],
        )
    except OSError as error:
        return {"unavailable": "the language server did not start: " + str(error)}
    server = Server(proc)
    try:
        return await ask(server, argv)
    finally:
        await server.close()


if __name__ == "__main__":
    sys.stdout.write(json.dumps(asyncio.run(main(sys.argv[1:]))) + "\n")

File: typescript/packages/core/src/tools/lsp-driver.ts
// The in-sandbox LSP client lsp runs with `python3 -c`: it starts the
// image's language server, opens one file, asks one question and prints one JSON line:
// {"ok": <LSP result>} or {"unavailable": <why>}. Stdlib only, so any image with python3 and the
// server answers the same way. It answers the server's own requests with empty results.

/** Where the driver looks up a bare server command, and the server's whole PATH. */
export const LSP_SEARCH_PATH =
  "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin";

export const LSP_DRIVER: string = String.raw`
import json, os, queue, shutil, subprocess, sys, threading, time

a = json.loads(sys.argv[1])
# The driver runs with the tool env, which has no PATH: the default search (/bin:/usr/bin)
# would miss servers in /usr/local/bin, where pip and npm put them, and a launcher's
# "#!/usr/bin/env node" would find no node. So the server is looked up in, and runs with a PATH
# of, exactly these system dirs: never a host PATH. An absolute command runs as given.
SEARCH = "${LSP_SEARCH_PATH}"
a["command"][0] = shutil.which(a["command"][0], path=SEARCH) or a["command"][0]
server_env = dict(os.environ, PATH=SEARCH)

def out(obj):
    sys.stdout.write(json.dumps(obj) + "\n")
    sys.stdout.flush()

try:
    proc = subprocess.Popen(a["command"], stdin=subprocess.PIPE, stdout=subprocess.PIPE,
                            stderr=subprocess.DEVNULL, cwd=a["root"], env=server_env)
except OSError as e:
    out({"unavailable": "language server %s not started: %s" % (a["command"][0], e)})
    sys.exit(0)

inbox = queue.Queue()

def reader():
    f = proc.stdout
    while True:
        length = None
        while True:
            line = f.readline()
            if not line:
                inbox.put(None)
                return
            line = line.strip()
            if not line:
                break
            if line.lower().startswith(b"content-length:"):
                length = int(line.split(b":")[1])
        if length is None:
            continue
        inbox.put(json.loads(f.read(length)))

threading.Thread(target=reader, daemon=True).start()

def send(msg):
    msg["jsonrpc"] = "2.0"
    body = json.dumps(msg).encode()
    proc.stdin.write(b"Content-Length: %d\r\n\r\n" % len(body) + body)
    proc.stdin.flush()

def answer(msg):
    method = msg.get("method")
    if method == "workspace/configuration":
        send({"id": msg["id"], "result": [None for _ in msg["params"].get("items", [])]})
    else:
        send({"id": msg["id"], "result": None})

def wait(pred, seconds):
    end = time.time() + seconds
    while time.time() < end:
        try:
            msg = inbox.get(timeout=max(0.01, end - time.time()))
        except queue.Empty:
            return None
        if msg is None:
            return None
        if "method" in msg and "id" in msg:
            answer(msg)
            continue
        if pred(msg):
            return msg
    return None

def request(id_, method, params, seconds):
    send({"id": id_, "method": method, "params": params})
    return wait(lambda m: m.get("id") == id_ and "method" not in m, seconds)

root_uri = "file://" + a["root"]
uri = "file://" + a["path"]
init = request(1, "initialize", {"processId": os.getpid(), "rootUri": root_uri,
    "workspaceFolders": [{"uri": root_uri, "name": "workspace"}],
    "capabilities": {"textDocument": {"publishDiagnostics": {}, "hover": {"contentFormat": ["plaintext"]},
                     "documentSymbol": {"hierarchicalDocumentSymbolSupport": True}}}}, a["ready_s"])
if init is None or "error" in init:
    out({"unavailable": "language server %s is not ready" % a["command"][0]})
    proc.kill()
    sys.exit(0)
send({"method": "initialized", "params": {}})
with open(a["path"], encoding="utf-8", errors="replace") as fh:
    text = fh.read()
send({"method": "textDocument/didOpen", "params": {"textDocument": {
    "uri": uri, "languageId": a["language_id"], "version": 1, "text": text}}})

op = a["operation"]
if op == "diagnostics":
    got = wait(lambda m: m.get("method") == "textDocument/publishDiagnostics"
               and m["params"]["uri"] == uri, a["wait_s"])
    if got is None:
        out({"unavailable": "no diagnostics were published for the file"})
    else:
        later = wait(lambda m: m.get("method") == "textDocument/publishDiagnostics"
                     and m["params"]["uri"] == uri, 1.0)
        out({"ok": (later or got)["params"]["diagnostics"]})
else:
    pos = {"line": a.get("line", 1) - 1, "character": a.get("character", 1) - 1}
    doc = {"textDocument": {"uri": uri}}
    method, params = {
        "definition": ("textDocument/definition", dict(doc, position=pos)),
        "references": ("textDocument/references", dict(doc, position=pos, context={"includeDeclaration": True})),
        "hover": ("textDocument/hover", dict(doc, position=pos)),
        "symbols": ("textDocument/documentSymbol", doc),
    }[op]
    got = request(2, method, params, a["wait_s"])
    if got is None:
        out({"unavailable": "the language server did not answer %s" % method})
    elif "error" in got:
        out({"unavailable": "the language server refused %s: %s" % (method, got["error"].get("message"))})
    else:
        out({"ok": got.get("result")})
try:
    request(3, "shutdown", None, 2)
    send({"method": "exit"})
except OSError:
    pass
proc.kill()
`;
